src/Executor.hs

```diff
@@ -63,35 +63,116 @@ data AdminInstr =
   | Trapping String {- Trap with error message -}
   | Returning (Stack WasmVal)
   | Breaking Int32 (Stack WasmVal)
-  -- | Label of int * instr list * code
+  | Label Int32 [Instr] Code
   | Frame Frame Code
   deriving (Show, Eq)
 
 
 data Config =
-  Config Frame Code
+  Config { confFrame :: Frame, confCode :: Code }
   deriving (Show, Eq)
 
 {- reduction -}
 
 step :: Config -> Config
-step (Config frame (Code vs es)) = do
-  let (vs', es') = case (head es, vs) of
-                      (Plain e', vs) ->
-                        case (e', vs) of
-                          (Nop, vs) ->
-                            (vs, [])
-
-                          {- TODO: factor Numeric (binary, unary etc. out into its own function/file-}
-                          (Numeric e', v2 : v1 : vs') ->
-                              case e' of
-                                (Add typ) -> (binOp typ v1 v2 (+) (+) : vs', [])
-                                _     -> error "unimplemented numeric"
-
-                      _ -> error "unimplemented instruction"
-
+step c@(Config frame (Code vs es)) = do
+  let (vs', es') = step' (head es) vs c
   Config frame (Code vs' (es' ++ tail es))
 
+-- |Intermediate step function
+-- takes admin_instr and a value stack
+-- returns the modified value stack
+step' :: AdminInstr -> Stack WasmVal -> Config -> (Stack WasmVal, [AdminInstr])
+
+-- Definitions for plain instructions
+step' (Plain e) vs _ = case (e, vs) of
+  (Nop, vs) ->
+    (vs, [])
+
+  (Unreachable, vs) ->
+    (vs, [Trapping "unreachable code executed"])
+
+  {- TODO: factor Numeric (binary, unary etc. out into its own function/file-}
+  (Numeric e', v2 : v1 : vs') ->
+    case e' of
+      (Add typ) -> (binOp typ v1 v2 (+) (+) : vs', [])
+      _     -> error "unimplemented numeric"
+
+  ((Bl inputTypes instructions), vs) -> do
```

Comment:
I believe inputTypes should be resultTypes instead. [spec](http://webassembly.github.io/spec/core/text/instructions.html#control-instructions)

Reply:
You're correct. Indeed inputTypes should be resultTypes.